### PR TITLE
Verify downloaded file's MD5

### LINKED ISSUE #93: Verify provided MD5 after download

### PR DESCRIPTION
**Pull request by jermnelson**
## Why was this change made?
To verify that the downloaded files in a Google book has the same MD5 hash as contained in the `checksum.md5` file in the download tar file.

Fixes #93 

## Was the documentation (README, DevOpsDocs, wiki, openapi.json) updated?
n/a

### CODE REVIEW

File: app/lib/grin/book_file.rb [hunk 1]
```diff
@@ -1,6 +1,9 @@
 # frozen_string_literal: true
 
+require 'digest'
+require 'gpgme'
```
Comment (jcoyne) [inline app/lib/grin/book_file.rb]: Is this used?  Can you point out where?

File: app/lib/grin/book_file.rb [hunk 2]
```diff
@@ -1,6 +1,9 @@
...
 require 'rubygems/package'
+require 'zlib'
```
Comment (jcoyne) [inline app/lib/grin/book_file.rb]: and this one?

File: app/lib/grin/book_file.rb (hunk 1 above)
Comment (jermnelson) [reply app/lib/grin/book_file.rb]: Removing, thought it was necessary of digest and decrypting.

File: app/lib/grin/book_file.rb (hunk 2 above)
Comment (jermnelson) [reply app/lib/grin/book_file.rb]: Removing, leftover cruft.

File: README.md [hunk 3]
```diff
@@ -21,6 +21,8 @@ using the [OAuth 2.0 for Installed Applications](https://developers.google.com/a
 2. Access token: To acquire an access token, execute `bundle exec rake grin:authorize_by_browser` or `bundle exec rake grin:authorize_by_code`.
    Authorize_by_browser is slightly easier, but will only work in an environment in which a browser can be opened. The tokens
    will be stored as `creds.json`.
+3.  To use the `gpgme` you may also need to run `brew install gpg` to install
```
Comment (jcoyne) [inline README.md]: Should this be removed too?

Comment (jermnelson) [reply README.md]: I needed to install `gpg` in order to decrypt the `.gpg` download file from Google. I'll revise to remove the `gpgme` reference.

File: app/lib/grin/book_file.rb [hunk 4]
```diff
@@ -123,5 +127,38 @@ def process_entry(entry, dest)
...
+        end
+        errors
+      end
+
+      private
+
+      def compare_md5(complete_filepath, md5_hash)
+        existing_md5 = Digest::MD5.new
+        File.open(complete_filepath) do |file_obj|
+          existing_md5 << file_obj.read
```
Comment (jcoyne) [inline app/lib/grin/book_file.rb]: Will this end up reading the whole file into memory? Is there a streaming approach we could use? Did you try `Digest::MD5.file()`?

Comment (jermnelson) [reply app/lib/grin/book_file.rb]: Fixed with latest commit.

File: app/lib/grin/book_file.rb [hunk 5]
```diff
@@ -123,5 +127,38 @@ def process_entry(entry, dest)
...
+          end
+        end
+        errors
+      end
+
+      private
+
+      def compare_md5(complete_filepath, md5_hash)
+        existing_md5 = Digest::MD5.new
+        File.foreach(complete_filepath) do |line|
```
Comment (jcoyne) [inline app/lib/grin/book_file.rb]: IO.foreach reads every line, but a binary file isn't line based, so this buffer could effectively be the whole file.  Did you consider `Digest::MD5.file()`?

Comment (jermnelson) [reply app/lib/grin/book_file.rb]: Haven't found any good documentation on the use of `DIgest::MD5.file` method as a streaming option for calculating the MD5 checksum.

Comment (jcoyne) [reply app/lib/grin/book_file.rb]: I find that looking in the source code can help: https://github.com/ruby/ruby/blob/master/spec/ruby/library/digest/md5/file_spec.rb#L18

Comment (jermnelson) [reply app/lib/grin/book_file.rb]: Done. Using `digest::MD5.file` method call in latest commit.

**Pull request merged**